### PR TITLE
Enable CSE on OSSA

### PR DESCRIPTION
**Pull request by meg-gupta**

### CODE REVIEW

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 1]
```diff
@@ -86,6 +88,15 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
   using hash_code = llvm::hash_code;
 
 public:
+  static SILValue stripCopyValueInst(SILValue val) {
+    auto *inst = val->getDefiningInstruction();
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CSE.cpp:92]: A nice little one-liner that does this is:

```
while (auto *cvi = dyn_cast_or_null<CopyValueInst>(val->getDefiningInstruction())) {
  val = cvi->getOperand();
}
```

**Comment by gottesmm** [comment]
I started perusing maybe I missed, do you have a bunch of tests with guaranteed values?

**Comment by meg-gupta** [comment]
@gottesmm I added some tests with guaranteed values now.

**Comment by meg-gupta** [comment]
@swift-ci test

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 2]
```diff
@@ -86,31 +88,45 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
   using hash_code = llvm::hash_code;
 
 public:
+  static SILValue stripCopyValueInst(SILValue val) {
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/CSE.cpp:91]: Also, any chance we could try to strip borrows at some point as well?

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 3]
```diff
@@ -474,22 +566,35 @@ class CSE {
...
   CSE(bool RunsOnHighLevelSil, SideEffectAnalysis *SEA,
-      SILOptFunctionBuilder &FuncBuilder)
+      SILOptFunctionBuilder &FuncBuilder,
+      SILOpenedArchetypesTracker &OpenedArchetypesTracker,
+      InstructionCloner &Cloner)
       : SEA(SEA), FuncBuilder(FuncBuilder),
-        RunsOnHighLevelSil(RunsOnHighLevelSil) {}
+        OpenedArchetypesTracker(OpenedArchetypesTracker), Cloner(Cloner),
+        RunsOnHighLevelSil(RunsOnHighLevelSil) {
+    auto &Builder = Cloner.getBuilder();
```
Comment (zoecarver) [inline lib/SILOptimizer/Transforms/CSE.cpp:585]: Nit: also no captial. Or maybe even: `Cloner.getBuilder().setOpenedArchetypesTracker...`

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 4]
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
   return false;
 }
 
+static void
+collectGuaranteedTransitiveUses(BeginBorrowInst *inst,
+                                SmallVectorImpl<SILInstruction *> &result) {
```
Comment (meg-gupta) [inline lib/SILOptimizer/Transforms/CSE.cpp:928]: @gottesmm Need your feedback on this. Do you think it'll be useful to move this to BorrowedOperand/Value ?

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 5]
```diff
@@ -859,6 +924,148 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
...
+                           hasLifetimeEnd = true;
+                           return true;
+                         }
+                         return false;
+                       });
+        }
+        if (needsBorrow) {
+          SILBuilderWithScope(end).createEndBorrow(end->getLoc(), borrow);
+        }
+        if (needsCopy && (!lastUser || !hasLifetimeEnd)) {
```
[the remark this answers is not in the document]
Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp:1064]: Added some comments to explain

[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/CSE.cpp:1064]: Is there a case where `lastUser == false` and `hasLifetimeEnd == true`? I think this check could be simplified.

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 6]
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
...
+      // lifetime ending instruction was not found at the frontier
+      for (auto *end : frontier) {
+        auto endIt = end->getIterator();
+        SILInstruction *lastUser =
+            (endIt == end->getParent()->begin()) ? nullptr : &*std::prev(endIt);
+        bool hasLifetimeEnd = false;
+        // Find if the lastUser of copy was ending its lifetime
+        if (lastUser) {
+          auto lastUserOps = lastUser->getAllOperands();
+          hasLifetimeEnd = std::find_if(
```
Comment (zoecarver) [inline lib/SILOptimizer/Transforms/CSE.cpp:1051]: Two things: first, I'm surprised this doesn't fail to compile because `std::find_if` returns an iterator. I think this should check `std::find_if(...) != lastUserOps.end()`. Better yet, you could use `std::any_of`. Second, you can use `llvm::find_if`, and then you can pass it a container directly instead of a set of iterators (this applies to all algorithms in this patch).

Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp]: yes, thanks! I fixed this instance with std::any_of now.

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 7]
```diff
@@ -859,6 +923,152 @@ static bool isLazyPropertyGetter(ApplyInst *ai) {
...
+      // lifetime ending instruction was not found at the frontier
+      for (auto *end : frontier) {
+        auto endIt = end->getIterator();
+        SILInstruction *lastUser =
+            (endIt == end->getParent()->begin()) ? nullptr : &*std::prev(endIt);
+        bool hasLifetimeEnd = false;
+        // Find if the lastUser of copy was ending its lifetime
+        if (lastUser) {
+          auto lastUserOps = lastUser->getAllOperands();
+          hasLifetimeEnd = std::any_of(
```
Comment (zoecarver) [inline lib/SILOptimizer/Transforms/CSE.cpp:1051]: Nit: you can use `llvm::any_of` and then you don't have to give it both iterators. Sorry for all the back and forth on this piece of code :P

Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp:1051]: No! I didn't know about the llvm:: variants. I'll go over everything and fix them.

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 8]
```diff
@@ -460,10 +507,10 @@ void OwnershipRAUWUtility::eliminateReborrowsOfRecursiveBorrows(
     SILValue value = borrowingOperand->get();
     auto *bi = cast<BranchInst>(borrowingOperand->getUser());
     SILBuilderWithScope reborrowBuilder(bi);
-    auto *innerCopy = reborrowBuilder.createCopyValue(bi->getLoc(), value);
-    auto *innerBorrow =
-        reborrowBuilder.createBeginBorrow(bi->getLoc(), innerCopy);
-    auto *outerEndBorrow = reborrowBuilder.createEndBorrow(bi->getLoc(), value);
+    SILLocation loc = RegularLocation(bi->getLoc().getSourceLoc());
```
Comment (zoecarver) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:510]: Nit: couldn't this just be `bi->getLoc()`?

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 9]
```diff
@@ -1284,22 +1372,32 @@ class SILCSE : public SILFunctionTransform {
     auto *SEA = PM->getAnalysis<SideEffectAnalysis>();
     SILOptFunctionBuilder FuncBuilder(*this);
 
-    CSE C(RunsOnHighLevelSil, SEA, FuncBuilder);
+    auto *func = getFunction();
+    SILOpenedArchetypesTracker OpenedArchetypesTracker(func);
```
Comment (zoecarver) [inline lib/SILOptimizer/Transforms/CSE.cpp:1376]: Nit: `camelCase` not `TitleCase`.

**Review by zoecarver — APPROVED** [review]
This lgtm now. Thanks for all your work on this :)

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 10]
```diff
@@ -267,21 +301,29 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
...
   hash_code visitApplyInst(ApplyInst *X) {
-    OperandValueArrayRef Operands(X->getAllOperands());
-    return llvm::hash_combine(X->getKind(), X->getCallee(),
-                              llvm::hash_combine_range(Operands.begin(),
-                                                       Operands.end()),
-                              X->hasSubstitutions());
+    auto Operands = X->getOperandValues();
+    SmallVector<SILValue, 4> OperandValues;
+    llvm::transform(
+        Operands, std::back_inserter(OperandValues),
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CSE.cpp:307]: One thought. Is it possible to create a range that does this transform and pass that to hash_combine instead? That way we do not intermediate arrays (and potential mallocs).

(Keep in mind I am just tagging this one, but I am talking about all of the places that this was done.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 11]
```diff
@@ -73,6 +73,10 @@ Optional<SILBasicBlock::iterator> swift::getInsertAfterPoint(SILValue val) {
   if (isa<SingleValueInstruction>(val)) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/InstOptUtils.cpp:73]: why not:

```
if (auto *inst = val->getDefiningInstruction()) {
  return std::next(inst->getIterator());
}
```

**Review by gottesmm — COMMENTED** [review]
Some additional comments. I need to read this locally before giving it my +1. I am doing that now.

**Comment by gottesmm** [comment]
@meg-gupta I was talking with Andy about this issue around ValueLifetimeAnalysis and returns. I think we are hacking around bugs here. We need a way to have a /real/ location given a terminator insert pt and value lifetime analysis should be smart enough about return inst.

File: lib/SILOptimizer/Transforms/CSE.cpp (hunk 10 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp:307]: I'll try to do this

Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp]: I added a transform function as an optional argument to SILInstruction::getOperandValues. It may be useful in other code as well.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 12]
```diff
@@ -201,9 +201,15 @@ CopyValueInst *OwnershipLifetimeExtender::copyAndExtendForNonLifetimeEndingRAUW(
   auto opRange = makeUserRange(range);
   ValueLifetimeAnalysis lifetimeAnalysis(copy, opRange);
   ValueLifetimeAnalysis::Frontier frontier;
-  bool result = lifetimeAnalysis.computeFrontier(
-      frontier, ValueLifetimeAnalysis::DontModifyCFG, &ctx.deBlocks);
-  assert(result);
+  bool noCriticalEdges = lifetimeAnalysis.computeFrontier(
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:204]: I think this is the wrong way to do this. Why is this needed?

Comment (gottesmm) [reply lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:204]: Specifically, if it is needed due to the terminator issue we spoke about before, I think that Andy's recent change adding UnownedInstantaneousUse makes that not necessary for unowned things

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 13]
```diff
@@ -261,8 +261,9 @@ OwnershipLifetimeExtender::copyBorrowAndExtendForRAUW(SILValue newValue,
   while (!frontier.empty()) {
     auto *insertPt = frontier.pop_back_val();
     SILBuilderWithScope frontierBuilder(insertPt);
-    auto *ebi = frontierBuilder.createEndBorrow(insertPt->getLoc(), borrow);
-    auto *dvi = frontierBuilder.createDestroyValue(insertPt->getLoc(), copy);
+    SILLocation loc = RegularLocation(insertPt->getLoc().getSourceLoc());
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:264]: My suggestion: lets be careful and just use an autogenerated loc. Otherwise sometimes we can get weird stepping behavior sometimes.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 14]
```diff
@@ -56,7 +57,34 @@ replaceAllUsesAndEraseInner(SingleValueInstruction *svi, SILValue newValue,
         user->eraseFromParent();
       continue;
     }
+
+    // Special handling for copy_value with opened archetypes
+    // When replacing a copy_value's operand with a newValue, if the newValue's
+    // opened archetype is different from copy_value's opened archetype, we have
+    // to clone the copy_value, so that we have the correct opened archetype on
+    // copy_value's result type.
+    bool needsCloning = false;
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:66]: What is causing this to be needed? The reason I am asking is rather than expand what this routine does, I wonder if we could instead put the burden in some way on a different part of the code. Just the context would be useful.

File: include/swift/SIL/SILInstruction.h [hunk 15]
```diff
@@ -527,8 +527,11 @@ class SILInstruction
 public:
   using OperandValueRange =
       OptionalTransformRange<ArrayRef<Operand>, OperandToValue>;
-  OperandValueRange
-  getOperandValues(bool skipTypeDependentOperands = false) const;
+  OperandValueRange getOperandValues(
```
Comment (gottesmm) [inline include/swift/SIL/SILInstruction.h:530]: I am hesitant about this part of the change. My gut feeling says that the right thing to do here is to not put this upon all users of this code. Maybe the compiler will optimize it all away... I don't know. Why not instead add a composition type: make a transform range with this property at your local use site.

File: include/swift/SIL/InstructionUtils.h [hunk 16]
```diff
@@ -37,9 +37,13 @@ SILValue stripCasts(SILValue V);
...
 
-/// Return the underlying SILValue after stripping off all copy_value and
+/// Return the underlying SILValue after looking through all copy_value and
 /// begin_borrow instructions.
-SILValue stripOwnershipInsts(SILValue v);
+SILValue lookThroughOwnershipInsts(SILValue v);
+
+/// Return the underlying SILValue after looking through all copy_value
+/// instructions.
+SILValue lookThroughCopyValueInsts(SILValue v);
```
Comment (gottesmm) [inline include/swift/SIL/InstructionUtils.h:46]: Can you split out the lookThrough rename into a separate commit? It isn't really part of this commit (I am fine with it in this PR).

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 17]
```diff
@@ -182,53 +196,70 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
   hash_code visitStructInst(StructInst *X) {
     // This is safe since we are hashing the operands using the actual pointer
     // values of the values being used by the operand.
-    OperandValueArrayRef Operands(X->getAllOperands());
-    return llvm::hash_combine(X->getKind(), X->getStructDecl(),
-      llvm::hash_combine_range(Operands.begin(), Operands.end()));
+    auto OperandValues =
+        X->getFunction()->hasOwnership()
+            ? X->getOperandValues(false, lookThroughOwnershipInsts)
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CSE.cpp:201]: Why not create a local transform kind here? Why should the needs of CSE impact all use sites of operand values?

**Review by gottesmm — COMMENTED** [review]
More initial comments.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp (hunk 12 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:204]: We may have to split the edges while computing frontiers if we have a use on the terminator instruction.

Here's an example:

```
sil [ossa] @struct_extract_test4 : $@convention(thin) (@guaranteed StructWithEnum2) -> () {
bb0(%0 : @guaranteed $StructWithEnum2):
 %1 = struct_extract %0 : $StructWithEnum2, #StructWithEnum2.val
 %copy1 = copy_value %1 : $FakeOptional2
 br bb4

bb4:
 %2 = struct_extract %0 : $StructWithEnum2, #StructWithEnum2.val
 %copy2 = copy_value %2 : $FakeOptional2
 switch_enum %2 : $FakeOptional2, case #FakeOptional2.some1!enumelt:bb5, case #FakeOptional2.some2!enumelt:bb6

bb5(%arg4 : $UInt):
 br bb7(%arg4 : $UInt) --> needs split here

bb6(%arg5 : @guaranteed $Klass):
 %4 = unchecked_trivial_bit_cast %arg5 : $Klass to $UInt
 br bb7(%4 : $UInt)

bb7(%arg6 : $UInt):
 destroy_value %copy2 : $FakeOptional2
 br bb8

bb8:
 destroy_value %copy1 : $FakeOptional2
 %res = tuple ()
 return %res : $()
}
```
Here we are CSE'ing the `struct_extract` at %2 with %1. For this we will have to split the edges at bb5 to get a frontier after the branch.

File: include/swift/SIL/SILInstruction.h (hunk 15 above)
Comment (meg-gupta) [reply include/swift/SIL/SILInstruction.h]: I added a new overload of `getOperandValues` without default args, and created a new `TransformedOperandValueRange`.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp (hunk 12 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Utils/OwnershipOptUtils.cpp]: You are right, this was not needed. It turned out to be a bug where the ownership rauw was considering transitive uses of none ownership as well.

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp (hunk 14 above)
Comment (meg-gupta) [reply lib/SILOptimizer/Utils/OwnershipOptUtils.cpp]: I removed handling open_existential_ref in CSE. I'll follow this up with another PR with spl handling needed to deal with type dependent operands.

File: lib/SILOptimizer/Utils/InstOptUtils.cpp [hunk 18]
```diff
@@ -70,8 +70,8 @@ const std::function<void(SingleValueInstruction *, SILValue)>
         };
 
 Optional<SILBasicBlock::iterator> swift::getInsertAfterPoint(SILValue val) {
-  if (isa<SingleValueInstruction>(val)) {
-    return std::next(cast<SingleValueInstruction>(val)->getIterator());
+  if (auto *inst = val->getDefiningInstruction()) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Utils/InstOptUtils.cpp:73]: Also, are you using this to insert instructions? Wouldn't SILBuilder::insertAfter work?

File: lib/SILOptimizer/Utils/OwnershipOptUtils.cpp [hunk 19]
```diff
@@ -407,6 +407,9 @@ static void getAllBorrowedValueUsePoints(
   for (unsigned i = firstOffset; i < usePoints.size(); ++i) {
     if (auto fOperand = ForwardingOperand::get(usePoints[i])) {
       fOperand->visitForwardedValues([&](SILValue transitiveValue) {
+        // Do not include transitive uses with 'none' ownership
+        if (transitiveValue.getOwnershipKind() == OwnershipKind::None)
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/OwnershipOptUtils.cpp:411]: Nice catch!

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 20]
```diff
@@ -917,19 +1015,24 @@ bool CSE::processNode(DominanceInfoNode *Node) {
...
-        // nextI must be reassigned.
-        nextI = std::next(Inst->getIterator());
-        Inst->replaceAllUsesPairwiseWith(AvailInst);
-        Inst->eraseFromParent();
+      if (!isa<OpenExistentialRefInst>(Inst) ||
+          processOpenExistentialRef(cast<OpenExistentialRefInst>(Inst),
+                                    cast<OpenExistentialRefInst>(AvailInst))) {
+        if (!Inst->getFunction()->hasOwnership() ||
+            Inst->getResults().empty()) {
+          Inst->replaceAllUsesPairwiseWith(AvailInst);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/CSE.cpp:1023]: Can you move no results to a separate case? Seems weird to me call replaceAllUsesPairwiseWith it since it has no results so no uses!

**Review by gottesmm — COMMENTED** [review]
Some more final comments! Looks pretty good! I think we are close!

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci benchmark

File: lib/SILOptimizer/Transforms/CSE.cpp [hunk 21]
```diff
@@ -267,21 +314,37 @@ class HashVisitor : public SILInstructionVisitor<HashVisitor, llvm::hash_code> {
   }
 
   hash_code visitApplyInst(ApplyInst *X) {
+    if (X->getFunction()->hasOwnership()) {
+      auto TransformedOperandValues =
+          X->getOperandValues(lookThroughOwnershipInsts, false);
+      return llvm::hash_combine(
+          X->getKind(), X->getCallee(),
+          llvm::hash_combine_range(TransformedOperandValues.begin(),
```
[the remark this answers is not in the document]
Comment (meg-gupta) [reply lib/SILOptimizer/Transforms/CSE.cpp:322]: I tried this, it doesn't take a range

**Comment by meg-gupta** [comment]
@swift-ci test

**Comment by meg-gupta** [comment]
@swift-ci test

File: include/swift/SIL/SILInstruction.h [hunk 22]
```diff
@@ -523,12 +523,20 @@ class SILInstruction
...
   using OperandValueRange =
       OptionalTransformRange<ArrayRef<Operand>, OperandToValue>;
+  using TransformedOperandValueRange =
+      OptionalTransformRange<ArrayRef<Operand>, OperandToTransformedValue>;
+
   OperandValueRange
   getOperandValues(bool skipTypeDependentOperands = false) const;
+  TransformedOperandValueRange
+  getOperandValues(std::function<SILValue(SILValue)> transformFn,
+                   bool skipTypeDependentOperands) const;
```
[the remark this answers is not in the document]
Comment (meg-gupta) [reply include/swift/SIL/SILInstruction.h:539]: Ran into an overload resolution issue with clang when I use a default value for `skipTypeDependentOperands`.
See https://godbolt.org/z/17514r. Leaving it as is for now.

**Comment by meg-gupta** [comment]
@gottesmm I am merging this. I'll add a follow on commit if there are any other comments.

File: include/swift/SIL/SILInstruction.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (zoecarver) [reply include/swift/SIL/SILInstruction.h:539]: I think the issue is that a pointer picks the implicit bool conversion over the converting constructor in `std::function`. I think it's fine the way it is, but one way to fix it would be to explicitly construct a `std::function` in the caller.

**Comment by zoecarver** [comment]
Woohoo! So excited to see ownership making its way through the SILOptimizer. Thanks for all your work on this, @meg-gupta.

File: include/swift/SIL/SILInstruction.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (meg-gupta) [reply include/swift/SIL/SILInstruction.h:539]: Yea, it can be a mild surprise. I didn't want to add it to avoid any confusion for new callers.
